server/src/main/java/org/apache/calcite/avatica/server/AvaticaJsonHandler.java

```diff
@@ -106,14 +106,18 @@ public void handle(String target, Request baseRequest,
           // Avoid a new buffer creation for every HTTP request
           final UnsynchronizedBuffer buffer = threadLocalBuffer.get();
           try (ServletInputStream inputStream = request.getInputStream()) {
-            rawRequest = AvaticaUtils.readFully(inputStream, buffer);
+            byte[] bytes = AvaticaUtils.readFullyToBytes(inputStream, buffer);
+            String encoding = request.getCharacterEncoding();
```

Comment:
`request#getReader()` might be better here, however `UnsynchronizedBuffer` is for `byte[]` only, and I didn't want to alter the code much.